R/roc.R

```diff
@@ -0,0 +1,261 @@
+#' out <- return_SE_SP(from = 0.00001, to = 0.999999999999999, by = 0.001, 
+#'                     pmix = c(1/4, 1/4, 1/4, 1/4), 
+#'                     alpha = list(0, -0.70, -1.05, -1.10), 
+#'                     psi = list(1, 1.2, 1.29, 1.3),
+#'                     nu = list(nu_matrix, nu_matrix1, nu_matrix2, nu_matrix3),
+#'                     lambda = list(lambda_matrix, lambda_matrix, lambda_matrix,
+#'                                   lambda_matrix),
+#'                     theta = list(theta_matrix, theta_matrix1, theta_matrix2,
+#'                                  theta_matrix3),
+#'                     plot_contour = FALSE, show_mi_result = TRUE, 
+#'                     mod_names = c("str", "par"))
+#' roc_auc_PartInv(out, plot_mods = c("strict", "partial"))
+#' roc_auc_PartInv(out, plot_mods = c("partial"))
+#' @export
+roc_auc_PartInv <- function(return_SE_SP_output, 
+                            plot_mods = c("partial", "strict"),
+                            plot_group = NULL, ...){
+  if(is.null(plot_group)){
+    plot_group <- c(1:length(rownames(out[1][[1]])))
+  }
+  
+  out <- return_SE_SP_output
+  labx <- "FPR (1-SP)"
+  laby <- "TPR (SE)"
+  SEs <- SPs <- data.frame()
+  AUCs <- c()
+  
+  par_ind <- grep("par", names(out))
+  str_ind <- grep("str", names(out))
+  # if partial invariance is indicated and `out` does contain partial inv. results
+  if("partial" %in% plot_mods && (!(is.integer(par_ind) && length(par_ind) == 0L))) {
+    partial <- c()
+    # for each group, extract SE and SP values, pad with 0 and 1, plot ROC
+    for(g in plot_group){
+      SEs <- out[par_ind[1]][[1]][g,]
+      SPs <- out[par_ind[2]][[1]][g,]
+      labg <- ifelse(g == 1, "Reference", paste0("Group ", g))
+      plot(y = c(0, SEs, 1), x = c(0, 1 - SPs,1), type = "l",
+           ylim = c(0,1), xlim = c(0,1), xlab = labx, ylab = laby,
+           main = paste0("ROC: ", labg," (Partial Invariance)"))
+      # compute AUC for the group 
+      partial <- cbind(partial, auc(SEs, SPs))
+    }
+    colnames(partial) <- paste0("Group ", plot_group)
+    AUCs[[grep("partial", plot_mods)]] <- partial
+  }
+  # if strict invariance is indicated and `out` does contain strict inv. results
+  if("strict" %in% plot_mods && (!(is.integer(str_ind) && length(str_ind) == 0L))) {
+    strict <- c()
+    # for each group, extract SE and SP values, pad with 0 and 1, plot ROC
+    for(g in plot_group){
+      SEs <- out[str_ind[1]][[1]][g,]
+      SPs <- out[str_ind[2]][[1]][g,]
+      labg <- ifelse(g == 1, "Reference", paste0("Group ", g))
+      plot(y = c(0, SEs, 1), x = c(0, 1 - SPs,1), type = "l",
+           ylim = c(0,1), xlim = c(0,1), xlab = labx, ylab = laby,
+           main = paste0("ROC: ", labg," (Strict Invariance)"))
+      # compute AUC for the group
+      strict <- cbind(strict, auc(SEs, SPs))
+    }
+    colnames(strict) <- paste0("Group ", plot_group)
+    AUCs[[grep("strict", plot_mods)]] <- strict
+  }
```

Comment:
These two parts are basically the same, so you should be able to write one function and call it twice.